src/execution/operator/join/physical_hash_join.cpp

```diff
@@ -19,24 +19,57 @@ namespace duckdb {
 
 PhysicalHashJoin::PhysicalHashJoin(LogicalOperator &op, unique_ptr<PhysicalOperator> left,
                                    unique_ptr<PhysicalOperator> right, vector<JoinCondition> cond, JoinType join_type,
-                                   const vector<idx_t> &left_projection_map,
-                                   const vector<idx_t> &right_projection_map_p, vector<LogicalType> delim_types,
-                                   idx_t estimated_cardinality, PerfectHashJoinStats perfect_join_stats)
+                                   const vector<idx_t> &left_projection_map, const vector<idx_t> &right_projection_map,
+                                   vector<LogicalType> delim_types, idx_t estimated_cardinality,
+                                   PerfectHashJoinStats perfect_join_stats)
     : PhysicalComparisonJoin(op, PhysicalOperatorType::HASH_JOIN, std::move(cond), join_type, estimated_cardinality),
-      right_projection_map(right_projection_map_p), delim_types(std::move(delim_types)),
-      perfect_join_statistics(std::move(perfect_join_stats)) {
+      delim_types(std::move(delim_types)), perfect_join_statistics(std::move(perfect_join_stats)) {
+	D_ASSERT(left_projection_map.empty());
 
 	children.push_back(std::move(left));
 	children.push_back(std::move(right));
 
-	D_ASSERT(left_projection_map.empty());
-	for (auto &condition : conditions) {
+	// Collect condition types, and which conditions are just references (so we won't duplicate them in the payload)
+	unordered_map<idx_t, idx_t> build_columns_in_conditions;
+	for (idx_t cond_idx = 0; cond_idx < conditions.size(); cond_idx++) {
+		auto &condition = conditions[cond_idx];
 		condition_types.push_back(condition.left->return_type);
+		if (condition.right->GetExpressionClass() == ExpressionClass::BOUND_REF) {
+			build_columns_in_conditions.emplace(condition.right->Cast<BoundReferenceExpression>().index, cond_idx);
+		}
+	}
+
+	// For ANTI, SEMI and MARK join, we only need to store the keys, so for these the payload/RHS types are empty
+	if (join_type == JoinType::ANTI || join_type == JoinType::SEMI || join_type == JoinType::MARK) {
+		return;
```

Comment:
could this go before the loop that's on line 34? Seems like if we return here the work in the loop doesn't do anything.

Reply:
It can't. The loop fills `condition_types`, so it needs to be run even for `ANTI`/`SEMI`/`MARK`.